test/lit/passes/heap2local.wast

```diff
@@ -2216,3 +2242,903 @@
+          (call $get-i32)
+          (i32.const 1)
+        )
+        (i32.const 0)
+      )
+    )
+    (drop
+      (array.get $array
+        (array.new $array
+          (call $get-i32)
+          (i32.const 19)
+        )
+        (i32.const 6)
+      )
+    )
+
+    ;; 20 is too much, however.
+    (drop
+      (array.get $array
+        (array.new $array
+          (call $get-i32)
+          (i32.const 20)
+        )
+        (i32.const 6)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $array.nested.refinalize.get (type $1) (result i32)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result nullref)
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $array.nested.refinalize.get (result i32)
+    ;; The get here is of an index that is out of bounds, and will trap. We
+    ;; should refinalize so the unreachability is propagated and we do not
+    ;; error on validation.
+    (array.get $array
+      (array.new_default $array
+        (i32.const 0)
+      )
+      (i32.const 0)
+    )
+  )
+
+  ;; CHECK:      (func $array.nested.refinalize.set (type $3)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result nullref)
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (i32.const 42)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $array.nested.refinalize.set
+    ;; As above, but with a set.
+    (array.set $array
+      (array.new_default $array
+        (i32.const 0)
+      )
+      (i32.const 0)
+      (i32.const 42)
+    )
+  )
+
+  ;; CHECK:      (func $array.flowing.type (type $1) (result i32)
+  ;; CHECK-NEXT:  (local $temp (ref $array))
+  ;; CHECK-NEXT:  (local $1 i32)
+  ;; CHECK-NEXT:  (local $2 i32)
+  ;; CHECK-NEXT:  (local $3 i32)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result (ref null $2))
+  ;; CHECK-NEXT:    (local.set $1
+  ;; CHECK-NEXT:     (i32.const 1)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (block (result nullref)
+  ;; CHECK-NEXT:     (local.set $3
+  ;; CHECK-NEXT:      (local.get $1)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (local.set $2
+  ;; CHECK-NEXT:      (local.get $3)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (ref.null none)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result (ref null $2))
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result (ref null $2))
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (block (result i32)
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (ref.null none)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (local.get $2)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $array.flowing.type (result i32)
+    (local $temp (ref $array))
```

Comment:
Does this still work properly when the local is a supertype of the allocated array type?

Reply:
Yes. We had a test for that with structs and I added one for arrays now, good idea.